### PR TITLE
Make Clover tasks compatible with the build cache

### PR DESCRIPTION
**Pull request by ghale**
These changes make the Clover tasks compatible with the Gradle build cache.  There are a few breaking changes, so this would probably require a major version bump.  The following are the most impactful changes:

- These changes drop support for Gradle 2.x and 3.x.  Only Gradle 4.x, 5.x and 6.x are now supported.
- Code instrumentation now occurs in a separate task and writes instrumented class files to a separate directory.  The test task now removes the original class directory from its classpath and adds the instrumented class directory to its classpath.  There are various changes to `CloverSourceSet` to support this.  Specifically, `CloverSourceSet.classesDir` no longer represents _both_ the original class dir and the instrumented class dir.  `classesDir` now represents the original (non-instrumented) class directory and `CloverSourceSet.instrumentedClassesDir` represents the instrumented class directory.
- In previous versions of this plugin, if `cloverInstrumentedJar` is set, any jar file that included the main sourceset output would automatically add the instrumented classes (because these classes overwrote the original class files).  Now, we automatically configure the `jar` task from the java plugin, but any other custom jar tasks would have to explicitly add the instrumented classes dir if they want them.
- A big part of this work is to eliminate the overlapping outputs between tasks as this is not compatible with the build cache.  For the most part, these overlapping outputs are somewhat transparent to the user, but one place where we can't avoid it is with the reports directory in multiproject builds.  In previous versions, merging the aggregate reports would overwrite any reports written in the report directory of the root project.  With these changes, we write the merged aggregate reports to a different location than the reports from the root project.  The reports from the root project itself still go into `build/reports/clover` while the merged aggregate reports now go into `build/reports-all/clover` in the root project.
- The same is true of the merged aggregate Clover database.  Instead of overwriting the Clover database in the root project, it now writes the merged database to `build/.clover/clover.db-all`.

### CODE REVIEW

**Comment by Alex-Vol-SV** [comment]
Wow @ghale you did not disappoint! This is a fantastic change. In fact it probably takes care of the issues I was already aware of including the problem that a test task might do the swapping of classes in a multiproject build while another java sibling project is attempting to compile against the original classes.

It will take me some time to digest your changes but the overall approach is very similar to the conclusions I had made as I was pondering the solution to the problems.

File: src/main/groovy/com/bmuschko/gradle/clover/CloverPlugin.groovy [hunk 1]
```diff
@@ -105,17 +98,27 @@ class CloverPlugin implements Plugin<Project> {
...
+        return "cloverInstrumentCodeFor${testTask.name.capitalize()}"
+    }
+
     @CompileStatic
     private void configureActionsForTask(Test test, Project project, CloverPluginConvention cloverPluginConvention, SourceSetsResolver resolver, AggregateDatabasesTask aggregateDatabasesTask) {
         if (testTaskEnabled(test, cloverPluginConvention)) {
+            // Add instrumentation task
+            def instrumentCodeTask = project.tasks.create(getInstrumentationTaskName(test), CloverInstrumentationTask, cloverPluginConvention, test, resolver)
+            instrumentCodeTask.dependsOn(test.testClassesDirs)
+            test.dependsOn(instrumentCodeTask)
```
Comment (Alex-Vol-SV) [inline src/main/groovy/com/bmuschko/gradle/clover/CloverPlugin.groovy]: This is very elegant solution to the problem @ghale

File: src/main/groovy/com/bmuschko/gradle/clover/CloverPlugin.groovy [hunk 2]
```diff
@@ -122,29 +131,23 @@ class CloverPlugin implements Plugin<Project> {
             test.include optimizeTestSetAction // action is also a file inclusion spec
             test.doLast createCreateSnapshotAction(cloverPluginConvention, project, test)
             if (project.hasProperty('cloverInstrumentedJar')) {
+
                 log.info "Skipping RestoreOriginalClassesAction for {} to generate instrumented JAR", test
```
Comment (Alex-Vol-SV) [inline src/main/groovy/com/bmuschko/gradle/clover/CloverPlugin.groovy]: This log statement is no longer true. Delete it?

File: build.gradle [hunk 3]
```diff
@@ -77,13 +70,18 @@ ext.compatibilityVersion = '1.8'
...
-jar {
-    manifest {
-        attributes 'Implementation-Title': 'Gradle Clover plugin',
-                   'Implementation-Version': version,
-                   'Built-By': System.getProperty("user.name"),
-                   'Built-Date': new Date(),
-                   'Built-JDK': System.getProperty("java.version")
+// Only add changing data to the manifest when publishing
+gradle.taskGraph.whenReady { graph ->
+    if (graph.allTasks.any { task -> task instanceof PublishToMavenLocal || task instanceof PublishToMavenRepository }) {
```
Comment (Alex-Vol-SV) [inline build.gradle:75]: What about use cases where Ivy is used? I use Ivy for artifacts in my use cases at work.

File: src/main/groovy/com/bmuschko/gradle/clover/CloverPlugin.groovy (hunk 2 above)
Comment (ghale) [reply src/main/groovy/com/bmuschko/gradle/clover/CloverPlugin.groovy]: Yes, this was eventually removed by df90bd7f9abbd8c7b202ed108755fccf6900cafc

File: build.gradle (hunk 3 above)
Comment (ghale) [reply build.gradle:75]: So this is just in the gradle-clover-plugin build.  If you are publishing this plugin to Ivy, then yeah, we should probably address those as well.  I just didn't see any evidence of Ivy publishing configuration in the build (in build.gradle or gradle/publishing.gradle).

Comment (Alex-Vol-SV) [reply build.gradle:75]: Ah, you are right. Never mind then. My mistake.

**Comment by Alex-Vol-SV** [comment]
@ghale if you have no further updates I will squash the commits and merge to master in preparation for release. I have some additional changes I was working on that I plan to add before I publish the 3.0.0 release with this. I will update the README/RELEASE_NOTES with the new features/capabilities after I confirm what other open issues are fixed with this.

**Comment by ghale** [comment]
Yes, I'm done with updates unless any other issues surface.

**Comment by Alex-Vol-SV** [comment]
Thanks @ghale for this great change. I appreciate the work that went into it.
I plan to have this in the 3.0.0 release by the weekend if not sooner.

**Pull request merged**